### PR TITLE
Fixes integer overflow in mask_iou (fast, variant 3)

### PR DESCRIPTION
**Pull request by Alarmod**
fix: mask_iou now doesn't have integer overflow if you pass a set of uint8 masks. The call to torch.matmul has been replaced by a combined call to torch.sum and torch.mul.

Test: 
```python
import torch

from ultralytics.utils.metrics import mask_iou

def mask_iou_mod(mask1, mask2, eps=1e-7):
    intersection = torch.matmul(mask1.to(torch.float32), mask2.to(torch.float32).T).clamp_(0)
    union = (mask1.sum(1)[:, None] + mask2.sum(1)[None]) - intersection  # (area1 + area2) - intersection
    return intersection / (union + eps)

def mask_iou_alarmod(mask1, mask2, eps=1e-7):
    intersection = torch.sum(torch.mul(mask1[:,None], mask2),dim=-1)
    union = (mask1.sum(1)[:, None] + mask2.sum(1)[None]) - intersection  # (area1 + area2) - intersection
    return intersection / (union + eps)

mask1_a = torch.randint(low=0, high=2, size=(1, 10000)).to(torch.uint8)
mask1_b = torch.randint(low=0, high=2, size=(1, 10000)).to(torch.uint8)
mask1 = torch.cat((mask1_a, mask1_b))

mask2_a = torch.randint(low=0, high=2, size=(1, 10000)).to(torch.uint8)
mask2_b = torch.randint(low=0, high=2, size=(1, 10000)).to(torch.uint8)
mask2_c = torch.randint(low=0, high=2, size=(1, 10000)).to(torch.uint8)
mask2 = torch.cat((mask2_a, mask2_b, mask2_c))

iou_original = mask_iou(mask1, mask2)
print(iou_original)

iou_mod = mask_iou_mod(mask1, mask2)
print(iou_mod)

iou_alarmod = mask_iou_alarmod(mask1, mask2)
print(iou_alarmod)
```

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Improved the calculation method for mask Intersection over Union (IoU) to enhance accuracy and reliability. 🖼️✨

### 📊 Key Changes  
- Updated the mask IoU calculation to use element-wise multiplication and summation instead of matrix multiplication.
- Ensured the intersection between masks is computed more precisely.

### 🎯 Purpose & Impact  
- Provides more accurate evaluation metrics for segmentation tasks.
- Improves the reliability of model performance assessments, especially for users working with object masks.
- Enhances the overall quality of results in segmentation workflows.

### CODE REVIEW

**Comment by Alarmod** [comment]
torch.sum work good long time (from version 1.1):  
torch_sum

It use internal upcast.

**Comment by glenn-jocher** [comment]
@Alarmod thanks for the PR! Do we need the clamp(0) operation here?

**Comment by Alarmod** [comment]
clamp_(0), as I understand it, is not needed for new code... The intersection is computed here as a logical AND operation, result of element-wise multiplication always 0 or 1, and sum result always >= 0.

**Comment by glenn-jocher** [comment]
You're correct; the `clamp_(0)` operation is unnecessary in the new code since the intersection is computed as a logical AND operation. Thanks for pointing that out!
